Add AWS CDK & esbuild docs to serverless (#24097)
DataDog/documentation: content/en/serverless/guide/serverless_tracing_and_bundlers.md

    **buildConsumer.js**
    ```javascript
    const ddPlugin = require('dd-trace/esbuild')
    const esbuild = require('esbuild')
    
    esbuild.build({
      entryPoints: ['./functions/consumer.ts'],
      outfile: 'out/consumer/index.js',
      plugins: [ddPlugin],
      // Other esbuild configuration
      external: [
        // esbuild cannot bundle native modules
        '@datadog/native-metrics',
    
        // required if you use profiling
        '@datadog/pprof',
    
        // required if you use Datadog security features
        '@datadog/native-appsec',
        '@datadog/native-iast-taint-tracking',
        '@datadog/native-iast-rewriter',
    
        // required if you encounter graphql errors during the build step
        'graphql/language/visitor',
        'graphql/language/printer',
        'graphql/utilities',

>         '@aws-sdk/client-sqs'

Reviewer: Let's add another entry here that is something like
```
// if you are using the package, instead of the layer
'datadog-lambda-js'
```

This suggestion also applies to the other ones in this file

      console.error(err)
      process.exit(1)
    })
    ```

4. When defining your `NodeJsFunction` in the CDK, use the `Code.fromCustomCommand` function to specify the path to your custom `esbuild` file and an output folder. For each separate Lambda function, specify the individual `esbuild` file defined in step three. The output folder should match the folder of the `outfile` in your `esbuild` file.

    **lambdaFunction.ts**
    ```typescript
    // This path will likely be different for each individual Lambda function
    const pathToBuildFile = '../functions/buildConsumer.js';
    
    // Ensure the files for each Lambda function are generated into their own directory
    const pathToOutputFolder = '../out/consumer/';
    
    const code = Code.fromCustomCommand(
      pathToOutputFolder,
      ['node', pathToBuildFile],
    );
    
    const consumerLambdaFunction = new NodejsFunction(this, props.functionName, {
      runtime: Runtime.NODEJS_20_X,
      code: code,
      handler: 'index.handler',
      memorySize: 512,
      bundling: {
        platform: 'node',
        esbuildArgs: {
          "--bundle": "true"
        },
        target: 'node20'
      }
    });
    ```

> ## AWS CDK with JS

Reviewer: Let's switch order so AWS CDK appears first?

We can drop the `with JS` from the name, as these docs are entirely for NodeJS

If you deploy NodeJS Lambda functions using the `NodeJsFunction` construct, but aren't using `esbuild` or Typescript you can still use Datadog to observe your serverless applications.

1. Follow the installation instructions for Node.js and ensure the Datadog Lambda layer for Node.js is added to your Lambda function.
2. Remove `datadog-lambda-js` and `dd-trace` from your `package.json` and the build process, since they are already available in the Lambda runtime provided by the Datadog Lambda Layer.

> 3. Use the `NodejsFunction` construct in the CDK. Ensure you set the `entry` property to be the path to the `Javascript` file containing your Lambda function handler, the `depsLockFilePath` to the path to your lock file for the package manager you are using and the `bundling.commandHooks.beforeBundling` to ensure all dependencies are installed.

Reviewer: ```suggestion
3. Use the `NodejsFunction` construct in the CDK. Ensure you set the `entry` property to be the path to the file containing your Lambda function handler, the `depsLockFilePath` to the path to your lock file for the package manager you are using and the `bundling.commandHooks.beforeBundling` to ensure all dependencies are installed.
```

    module.exports = [nodeExternalsPlugin()]
    ```

## AWS CDK

> If you deploy NodeJS Lambda functions using the `NodeJsFunction` construct, but aren't using `esbuild` or Typescript you can still use Datadog to observe your serverless applications.

Reviewer 2: ```suggestion
If you deploy Node.js Lambda functions using the `NodeJsFunction` construct, but aren't using `esbuild` or Typescript, you can still use Datadog to observe your serverless applications.
```

1. Follow the installation instructions for Node.js and ensure the Datadog Lambda layer for Node.js is added to your Lambda function.
2. Remove `datadog-lambda-js` and `dd-trace` from your `package.json` and the build process, since they are already available in the Lambda runtime provided by the Datadog Lambda Layer.

> 3. Use the `NodejsFunction` construct in the CDK. Ensure you set the `entry` property to be the path to the file containing your Lambda function handler, the `depsLockFilePath` to the path to your lock file for the package manager you are using and the `bundling.commandHooks.beforeBundling` to ensure all dependencies are installed.

Reviewer 2: ```suggestion
3. Use the `NodejsFunction` construct in the CDK. Ensure you set the `entry` property to be the path to the file containing your Lambda function handler, the `depsLockFilePath` to the path to your lock file for the package manager you are using, and the `bundling.commandHooks.beforeBundling` to ensure all dependencies are installed.
```

    module.exports = [nodeExternalsPlugin()]
    ```

> ## AWS CDK & esbuild

Reviewer: Would it make sense to cover customers who are only using the CDK configuration without esbuild?

Author: Yeah, that's not a bad shout. I'll update that today.

Author: @duncanista I've added an additional section. Let me know if that looks right, I've tested the code and I still get all the tracing.

The `NodeJsFunction` construct in the AWS CDK uses esbuild. The default configuration is not compatible with Datadog's tracing libraries. The CDK allows you to override the default configuration and provide a custom esbuild file to support bundling and the Datadog tracing libraries:

> 1. Follow the installation instructions for Node.js and ensure the Datadog Lambda layer for Node.js is added to your Lambda function.

Reviewer: Would it make sense to also cover users who are packages instead of the layer?

Author: Just to confirm, you mean people who are bundling the dd-trace package inside their code and not adding the layer at all?

Reviewer: So the Layer contains both `datadog-lambda-js` and `dd-trace` package. Some customers might not add layers, but instead of the package, and do the instrumentation in a different way.

At the end of the day, they can pick any of the ones they want!

> 2. Remove datadog-lambda-js and dd-trace from your package.json and the build process, since they are already available in the Lambda runtime provided by the Datadog Lambda layer.

Reviewer: ```suggestion
2. Remove `datadog-lambda-js` and `dd-trace` from your package.json and the build process, since they are already available in the Lambda runtime provided by the Datadog Lambda Layer.
```

> 2. Remove datadog-lambda-js and dd-trace from your package.json and the build process, since they are already available in the Lambda runtime provided by the Datadog Lambda layer.

Reviewer 3: ```suggestion
2. Remove `datadog-lambda-js` and `dd-trace` from your `package.json` and the build process, since they are already available in the Lambda runtime provided by the Datadog Lambda layer.
```

3. Create an `esbuild` file for each of your Lambda functions. A seperate `esbuild` file is required per Lambda function so that each entry point can be specified seperately. Notice the `entryPoint` and `outfile` properties. For example, if you had a second Lambda function in your project named `producer`, then the entry point would be `./functions/producer.ts` and the outfile would be `/out/producer/index.js`

    **buildConsumer.js**

>     ```

Reviewer: Specify language so docs code is highlighted

    const ddPlugin = require('dd-trace/esbuild')
    const esbuild = require('esbuild')
    
    esbuild.build({
      entryPoints: ['./functions/consumer.ts'],
      outfile: 'out/consumer/index.js',
      plugins: [ddPlugin],
      // Other esbuild configuration
      external: [
        // esbuild cannot bundle native modules
        '@datadog/native-metrics',
    
        // required if you use profiling
        '@datadog/pprof',
    
        // required if you use Datadog security features
        '@datadog/native-appsec',
        '@datadog/native-iast-taint-tracking',
        '@datadog/native-iast-rewriter',
    
        // required if you encounter graphql errors during the build step
        'graphql/language/visitor',
        'graphql/language/printer',
        'graphql/utilities',
        '@aws-sdk/client-sqs'
      ]
    }).catch((err) => {
      console.error(err)
      process.exit(1)
    })
    ```

> 4. When defining your `NodeJsFunction` in the CDK, use the `Code.fromCustomCommand` function to specify the path to your custom esbuild file and an output folder. For each seperate Lambda function, specify the individual `esbuild` file defined in step 3. The output folder should match the folder of the `outfile` in your `esbuild` file.

Reviewer: ```suggestion
4. When defining your `NodeJsFunction` in the CDK, use the `Code.fromCustomCommand` function to specify the path to your custom esbuild file and an output folder. For each separate Lambda function, specify the individual `esbuild` file defined in step 3. The output folder should match the folder of the `outfile` in your `esbuild` file.
```

> 4. When defining your `NodeJsFunction` in the CDK, use the `Code.fromCustomCommand` function to specify the path to your custom esbuild file and an output folder. For each seperate Lambda function, specify the individual `esbuild` file defined in step 3. The output folder should match the folder of the `outfile` in your `esbuild` file.

Reviewer 3: ```suggestion
4. When defining your `NodeJsFunction` in the CDK, use the `Code.fromCustomCommand` function to specify the path to your custom `esbuild` file and an output folder. For each seperate Lambda function, specify the individual `esbuild` file defined in step three. The output folder should match the folder of the `outfile` in your `esbuild` file.
```

    **lambdaFunction.ts**

>     ```

Reviewer: Specify language here too for docs highlight
